### PR TITLE
Fix `Get-ItemProperty` to report non-terminating error for cast exception

### LINKED ISSUE #9552: get-itemproperty : Unable to cast object of type 'System.Int64' to type 'System.Int32' (Specified cast is not valid).
Netbeans 8.2 (64-bit) installer currently puts a 64 bit integer (1) in a dword in the registry under the name NoModify under HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\nbi-nb-base-8.2.0.0.201609300101.  They don't seem to be in any rush to fix it.  https://netbeans.org/bugzilla/show_bug.cgi?id=251538  In Regedit, it says "(invalid dword (32-bit) value)".  Get-itemproperty could recover from this a bit better.  It doesn't show the other values, and other keys included after don't get displayed.  This is even with -erroraction silentlycontinue.  The error message doesn't point to the offending key.  Unless you pipe an error variable through format-list * -force.  Example superuser thread about it:  https://superuser.com/questions/1352166/how-to-resolve-error-specified-cast-is-not-valid-error-during-installation.  Error in PS 5.1 looks like:  "Specified cast is not valid."

# Steps to reproduce

```powershell
# with 64-bit Netbeans IDE 8.2 installed
get-itemproperty hklm:\software\microsoft\windows\currentversion\uninstall\*
```

# Expected behavior

```none
# display properties of all subkeys
```

# Actual behavior

```none
# displays some key properties, then...
get-itemproperty : Unable to cast object of type 'System.Int64' to type 'System.Int32'.
At line:1 char:1
+ get-itemproperty hklm:\software\microsoft\windows\currentversion\unin ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [Get-ItemProperty], InvalidCastException
+ FullyQualifiedErrorId : System.InvalidCastException,Microsoft.PowerShell.Commands.GetItemPropertyCommand
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      6.1.2
PSEdition                      Core
GitCommitId                    6.1.2
OS                             Microsoft Windows 10.0.16299
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```
As a test, you can import this as a .reg file, and it will put a 64 bit integer into a 32 bit dword registry item.  (a 32-bit dword would be "NoModify"=dword:00000001)

```
Windows Registry Editor Version 5.00

[HKEY_CURRENT_USER\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\nbi-nb-base-8.2.0.0.201609300101]
"NoModify"=hex(4):01,00,00,00,00,00,00,00
```

### PR DESCRIPTION
**Pull request by ArmaanMcleod**
# PR Summary

Fixes #9552

Adds catch for `InvalidCastException` so non-terminating error gets reported instead of terminating error when listing subkeys with 64 bit integer in 32 bit DWORD.

## PR Context

Currently if you do this you get a terminating error for the cast and command is halted. With this exception handling we can now get non-terminating error and also suppress error with `-ErrorAction SilentlyContinue`. `Get-ItemProperty` will continue to enumerate subkeys even if an invalid one is found.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1 [hunk 1]
```diff
@@ -471,6 +471,39 @@ Describe "Extended Registry Provider Tests" -Tags @("Feature", "RequireAdminOnWi
...
+            # Below will import .reg file with 64 bit integer in 32 bit DWORD
+            $badRegistryContent = @"
+Windows Registry Editor Version 5.00
+
+[$registrySubkeyPath]
+"NoModify"=hex(4):01,00,00,00,00,00,00,00
+"@
+            $badRegistryPath = Join-Path -Path $TestDrive -ChildPath badreg.reg
+            $badRegistryContent | Set-Content -Path $badRegistryPath
+            reg import $badRegistryPath
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1:488]:
```suggestion
            reg.exe import $badRegistryPath
```

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@ArmaanMcleod We need to skip new test code on Windows.

**Comment by ArmaanMcleod** [comment]
> @ArmaanMcleod We need to skip new test code on Windows.

@iSazonov Strange. I saw `$PSDefaultParameterValues["it:skip"] = !$IsWindows` at the top of test file so thought it skip non-windows.

> @ArmaanMcleod We need to skip new test code on Windows.

@iSazonov Do you mean we should skip if not windows and powershell core? Makes sense if this is for new test it should only work for powershell core.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1 (hunk 1 above)
Comment (ArmaanMcleod) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1]: Fixed in 2956793d5

File: test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1 [hunk 2]
```diff
@@ -471,6 +471,44 @@ Describe "Extended Registry Provider Tests" -Tags @("Feature", "RequireAdminOnWi
...
+
+                $badRegistryPath = Join-Path -Path $TestDrive -ChildPath badreg.reg
+                $badRegistryContent | Set-Content -Path $badRegistryPath
+                reg.exe import $badRegistryPath
+
+                $registryProviderSubkeyPath = 'HKCU:\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\badreg'
+            }
+        }
+
+        It "Validate non-terminating error for cast" -Skip (!$IsWindows) {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1:496]: The skip is already on top of the file.
```suggestion
        It "Validate non-terminating error for cast" {
```

File: test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1 [hunk 3]
```diff
@@ -471,6 +471,44 @@ Describe "Extended Registry Provider Tests" -Tags @("Feature", "RequireAdminOnWi
...
+            Get-ItemProperty -Path $registryProviderSubkeyPath -ErrorVariable err -ErrorAction SilentlyContinue
+            $err | Should -HaveCount 1
+            $err[0].Exception | Should -BeOfType [System.InvalidCastException]
+        }
+
+        It "Validate terminating error for cast" -Skip (!$IsWindows) {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1:502]:
```suggestion
        It "Validate terminating error for cast" {
```

File: src/System.Management.Automation/namespaces/RegistryProvider.cs [hunk 4]
```diff
@@ -1825,8 +1825,19 @@ private void MoveRegistryItem(string path, string destination)
...
+                    propertyResults.Properties.Add(new PSNoteProperty(notePropertyName, key.GetValue(valueName)));
+                    valueAdded = true;
+                }
+                catch (InvalidCastException invalidCast)
+                {
+                    WriteError(new ErrorRecord(
+                        invalidCast,
+                        invalidCast.GetType().FullName,
+                        ErrorCategory.WriteError, 
+                        path));
```
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/RegistryProvider.cs:1839]: Could you please show screenshot with the error message?

Comment (ArmaanMcleod) [reply src/System.Management.Automation/namespaces/RegistryProvider.cs:1839]: It is just the normal `InvalidCastException` message. I was thinking if we should use a more descriptive RESX to indicate which sub key path had issues?

Comment (ArmaanMcleod) [reply src/System.Management.Automation/namespaces/RegistryProvider.cs:1839]: Both `SetProperty` and `NewProperty` handle cast exception the same way it seems.

Comment (iSazonov) [reply src/System.Management.Automation/namespaces/RegistryProvider.cs:1839]: I see the code follow the file pattern so I am ok.

**Review by iSazonov — APPROVED** [review]

File: src/System.Management.Automation/namespaces/RegistryProvider.cs [hunk 5]
```diff
@@ -1825,8 +1825,19 @@ public void GetProperty(
...
+                try
+                {
+                    propertyResults.Properties.Add(new PSNoteProperty(notePropertyName, key.GetValue(valueName)));
+                    valueAdded = true;
+                }
+                catch (InvalidCastException invalidCast)
+                {
+                    WriteError(new ErrorRecord(
+                        invalidCast,
+                        invalidCast.GetType().FullName,
```
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/RegistryProvider.cs:1837]: I suggest to use "InvalidCast" const - it is our common pattern.

File: src/System.Management.Automation/namespaces/RegistryProvider.cs [hunk 6]
```diff
@@ -1825,8 +1825,19 @@ public void GetProperty(
...
+                        ErrorCategory.WriteError,
```
Comment (iSazonov) [inline src/System.Management.Automation/namespaces/RegistryProvider.cs:1838]: We read a value. Is there better error category?

File: src/System.Management.Automation/namespaces/RegistryProvider.cs (hunk 5 above)
Comment (ArmaanMcleod) [reply src/System.Management.Automation/namespaces/RegistryProvider.cs:1837]: @iSazonov Do you mean replacing the Error ID with a constant?

File: src/System.Management.Automation/namespaces/RegistryProvider.cs (hunk 6 above)
Comment (ArmaanMcleod) [reply src/System.Management.Automation/namespaces/RegistryProvider.cs:1838]: @iSazonov Maybe `ErrorCategory.InvalidOperation` would be better? I guess I was just following the same pattern as other parts of code when dealing with InvalidCastException.

File: src/System.Management.Automation/namespaces/RegistryProvider.cs (hunk 5 above)
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/RegistryProvider.cs:1837]: Yes, you can find many such patterns in code.

Comment (iSazonov) [reply src/System.Management.Automation/namespaces/RegistryProvider.cs:1837]: I found a pattern you mentioned. I agree to follow existing pattern.

#Closed

File: src/System.Management.Automation/namespaces/RegistryProvider.cs (hunk 6 above)
Comment (iSazonov) [reply src/System.Management.Automation/namespaces/RegistryProvider.cs:1838]: I see  there is `ErrorCategory.ReadError` in the file.

Comment (ArmaanMcleod) [reply src/System.Management.Automation/namespaces/RegistryProvider.cs]: Fixed in 44205c999baad1507e931bd7744f81d7970c9c55

**Review by iSazonov — APPROVED** [review]

**Pull request merged**